### PR TITLE
AstroCalc: Fix RTS

### LINKED ISSUE #2990: moon culmination data problem
Hello, With the astronomical calculations (F10 of the Stellarium interface) I asked to see the culminations of the moon over a given period. I then downloaded the data to excel. The problem is the following: the culmination times and the corresponding information given (height az; observer/moon distance, magnitude) are not always correct. There are in particular duplicates (two culminations for a single day and (then or before in the spreadsheet, everything is shifted. Is this a handling error on my part? a bug in the application? Thank you for your answer !

_Originally posted by @ibaneoudo in https://github.com/Stellarium/stellarium/discussions/2961_

### PR DESCRIPTION
**Pull request by worachate001**
### Description
The problems are:
- Stellarium calculates RTS of the Moon by choosing the nearest instances. Sometimes, rise or transit or set time occurs on previous day or next day.
- There are time spikes in AstroCalc's RTS table which is a remaining issue in #1482 (near bottom of the page).

This PR will fix them by:
1. When select the Moon (or other objects). Info will show '(previous day)' or '(next day)' after the time to inform us that the calculated event doesn't happen on current day.
2. RTS table in AstroCalc will show the events that occur on the same date in the same row. This should make it easier to read/process in other applications.
3. Fixing above helps us eliminate time spikes issue.

Moon-RTS

Fixes #2990 and remaining issue in #1482

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] This change requires a documentation update

### How Has This Been Tested?

**Test Configuration**:
* Operating system: Manjaro 22.0.1
* Graphics Card: -

## Checklist:

- [x] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [ ] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (header file)
- [ ] I have updated the respective chapter in the Stellarium User Guide
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by alex-w** [comment]
Thank you very much for the fix!

I see sentences “previous day/next day” - why not “yesterday/tomorrow”?

**Comment by 10110111** [comment]
> I see sentences “previous day/next day” - why not “yesterday/tomorrow”?

It makes sense if current day is not the realtime "today", and not even a simulated "today".

**Comment by worachate001** [comment]
> Thank you very much for the fix!
> 
> I see sentences “previous day/next day” - why not “yesterday/tomorrow”?

This actually came across my mind at one point during development. But I though the same with @10110111 that it's mostly simulated time. Using yesterday/tomorrow may not make sense.

**Comment by alex-w** [comment]
> This actually came across my mind at one point during development. But I though the same with @10110111 that it's mostly simulated time. Using yesterday/tomorrow may not make sense.

OK, I got it

**Review by alex-w — APPROVED** [review]

**Comment by gzotti** [comment]
Many thanks for this. `StelObject::getInfoMap()` may be fixed with this update too. The `transit`, `rise `and `set `entries are QStrings which may have `(previous day)` or `(next day)` appended (I think we usually don't translate the InfoMap to make it easier for programs which are using it). I am not sure how to deal with the `{transit|rise|set}-dhr` entries, those are numerical. Allow hours >24 and <0? Or just keep as-is and let the user of the data find out by parsing the other string?

**Comment by worachate001** [comment]
> Many thanks for this. `StelObject::getInfoMap()` may be fixed with this update too. The `transit`, `rise `and `set `entries are QStrings which may have `(previous day)` or `(next day)` appended (I think we usually don't translate the InfoMap to make it easier for programs which are using it). I am not sure how to deal with the `{transit|rise|set}-dhr` entries, those are numerical. Allow hours >24 and <0? Or just keep as-is and let the user of the data find out by parsing the other string?

I'm thinking about another way that might be better. Create getRTSTimeOfDate() that returns RTS of current date (doing it inside getRTSTime() may be too complicated).

**Comment by gzotti** [comment]
OK. Just that the InfoMap always needs updates in line with the InfoString.

**Comment by worachate001** [comment]
I'm ditching the idea of displaying 'previous day' or 'next day'. It was proposed earlier because that is an easy way with minimal tweak in existing codes. But it becomes more complicated for StelObject::getInfoMap().

New update will show RTS that occur on current date only (RTS will change at midnight, not seem to change in unpredictable way during a day like before). It also makes more sense and in line with the way they show RTS in astronomical almanacs. Please check StelObject::getInfoMap() to see if it's okay with new changes because I'm not familiar with its structure.

**Comment by alex-w** [comment]
Hmm... I see serious regress of performance on the Mac when the Moon is selected

File: src/core/modules/Planet.cpp [hunk 1]
```diff
@@ -4912,15 +4912,13 @@ void Planet::setApparentMagnitudeAlgorithm(QString algorithm)
...
 // Also, the estimate h0 for the Moon in the literature is based on geocentric computation.
 // NOTE: Limitation for efficiency: If this is a planet moon from another planet, we compute RTS for the parent planet instead!
-Vec4d Planet::getRTSTime(const StelCore *core, const double altitude) const
+Vec4d Planet::getClosestRTSTime(const StelCore *core, const double altitude) const
 {
 	const StelLocation loc=core->getCurrentLocation();
 	if (loc.name.contains("->")) // a spaceship
 		return Vec4d(0., 0., 0., -1000.);
 
-	// Keep time in sync (method from line 592) to fix slow down of time when the moon is selected
```
Comment (alex-w) [inline src/core/modules/Planet.cpp:4921]: Please back this fix

File: src/core/modules/Planet.cpp [hunk 2]
```diff
@@ -5051,7 +5049,6 @@ Vec4d Planet::getRTSTime(const StelCore *core, const double altitude) const
 			ms += ms2;
 		}
 		core1->setJD(currentJD);
-		core1->setMilliSecondsOfLastJDUpdate(millis); // restore millis.
```
Comment (alex-w) [inline src/core/modules/Planet.cpp:5054]: Second part for performance fix

**Comment by worachate001** [comment]
Thanks @alex-w. I have no performance hit so I'm not sure it's getting better or not. The new update should work now for all objects.

**Review by alex-w — APPROVED** [review]

**Pull request merged**